overlord/hookstate/ctlcmd/helpers.go

```diff
@@ -238,6 +244,114 @@ func runServiceCommand(context *hookstate.Context, inst *servicestate.Instructio
 	}
 }
 
+func validateSnapAndCompsNames(names []string, ctxSnap string) ([]string, error) {
+	var allComps []string
+	for _, name := range names {
+		snap, comps := snap.SplitSnapInstanceAndComponents(name)
+		// if snap is present it must be the context snap for the moment
+		if snap != "" && snap != ctxSnap {
+			return nil, fmt.Errorf("cannot install snaps using snapctl")
+		}
+		for _, comp := range comps {
+			if err := naming.ValidateSnap(comp); err != nil {
+				return nil, err
+			}
+		}
+		allComps = append(allComps, comps...)
+	}
+	return allComps, nil
+}
+
+type managementCommandTp int
+
+const (
+	installManagementCommand managementCommandTp = iota
+	removeManagementCommand
+)
+
+type managementCommand struct {
+	typ        managementCommandTp
+	components []string
+}
+
+func changeIdIfEphemeral(hctx *hookstate.Context) string {
+	if !hctx.IsEphemeral() {
+		return hctx.ChangeID()
+	}
+	return ""
+}
+
+func createSnapctlInstallTasks(hctx *hookstate.Context, cmd *managementCommand) (tss []*state.TaskSet, err error) {
+	st := hctx.State()
+	st.Lock()
+	defer st.Unlock()
+
+	info, err := currentSnapInfo(st, hctx.InstanceName())
+	if err != nil {
+		return nil, err
+	}
+	return snapstateInstallComponents(context.TODO(), st, cmd.components, info,
+		snapstate.Options{ExpectOneSnap: true, FromChange: changeIdIfEphemeral(hctx)})
+}
+
+func createSnapctlRemoveTasks(hctx *hookstate.Context, cmd *managementCommand) (tss []*state.TaskSet, err error) {
+	st := hctx.State()
+	st.Lock()
+	defer st.Unlock()
+
+	return snapstateRemoveComponents(st, hctx.InstanceName(), cmd.components,
+		snapstate.RemoveComponentsOpts{RefreshProfile: true,
+			FromChange: changeIdIfEphemeral(hctx)})
+}
+
+func runSnapManagementCommand(hctx *hookstate.Context, cmd *managementCommand) error {
```

Comment:
Since `cmd` isn't an optional parameter, and it is small, I think it would be simpler if it were accepted by-value, rather than as a pointer.

Reply:
Right, I've changed this in some functions now